### ISSUE TITLE
attempt to write a readonly database (code 1032, errno 0)

### ISSUE THREAD

**Issue by dotw** [Tencent/wcdb#193]
### 对应的 WCDB 语言 ( The language of WCDB )
Java 
### WCDB 的版本号 ( The version of WCDB )

v1.0.5
### WCDB 的平台 ( The platform of WCDB )
 Android
### WCDB 的安装方式 ( The installation of WCDB )
 AAR Package
### 遇到的问题 ( What's the issue? )
can't write database.

Steps:
1. copy an exist sqlcipher database to the database folder in app
2. open database with the correct password
3. keep the SQLiteDatabase as reference for future reuse.
4. jump to another screen to use the SQLiteDatabase object to write a record.
5. error is occurred.

full stack is bellow.
```
12-04 04:40:43.281 12026-12187/com.hzqi.sango E/MtaSDK: com.tencent.wcdb.database.SQLiteReadOnlyDatabaseException: attempt to write a readonly database (code 1032, errno 0): 
                                                            at com.tencent.wcdb.database.SQLiteConnection.nativeExecuteForChangedRowCount(Native Method)
                                                            at com.tencent.wcdb.database.SQLiteConnection.executeForChangedRowCount(SQLiteConnection.java:800)
                                                            at com.tencent.wcdb.database.SQLiteSession.executeForChangedRowCount(SQLiteSession.java:709)
                                                            at com.tencent.wcdb.database.SQLiteStatement.executeUpdateDelete(SQLiteStatement.java:91)
                                                            at com.tencent.wcdb.database.SQLiteDatabase.executeSql(SQLiteDatabase.java:1920)
                                                            at com.tencent.wcdb.database.SQLiteDatabase.execSQL(SQLiteDatabase.java:1871)
                                                            at com.hzqi.sango.android.database.GameManager.add(GameManager.java:40)
                                                            at com.hzqi.sango.entity.Game.newGame(Game.java:215)
                                                            at com.hzqi.sango.entity.type.ScreenEnum$3.getScreen(ScreenEnum.java:54)
                                                            at com.hzqi.sango.manager.SceneManager.showScreen(SceneManager.java:260)
                                                            at com.hzqi.sango.screen.ChooseRoleScreen.startGame(ChooseRoleScreen.java:281)
                                                            at com.hzqi.sango.screen.ChooseRoleScreen.access$500(ChooseRoleScreen.java:34)
                                                            at com.hzqi.sango.screen.ChooseRoleScreen$2.clicked(ChooseRoleScreen.java:144)
                                                            at com.badlogic.gdx.scenes.scene2d.utils.ClickListener.touchUp(ClickListener.java:89)
                                                            at com.badlogic.gdx.scenes.scene2d.InputListener.handle(InputListener.java:59)
                                                            at com.badlogic.gdx.scenes.scene2d.Stage.touchUp(Stage.java:351)
                                                            at com.badlogic.gdx.InputMultiplexer.touchUp(InputMultiplexer.java:96)
                                                            at com.badlogic.gdx.backends.android.AndroidInput.processEvents(AndroidInput.java:383)
                                                            at com.badlogic.gdx.backends.android.AndroidGraphics.onDrawFrame(AndroidGraphics.java:493)
                                                            at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1553)
                                                            at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1253)
```

- [ x] 我已阅读 [wiki的文档](https://github.com/Tencent/wcdb/wiki) 和 [常见问题](https://github.com/Tencent/wcdb/wiki/常见问题)。

- [ x]  I've read and understood the [wiki](https://github.com/Tencent/wcdb/wiki/iOS-macOS-Tutorial).

**Comment by dotw** [comment]
I did more investigation with this issue. I have changed the sample-encrypted to produce the error but without any lucky. A new error is produced as following...
```
               --------- beginning of crash
E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #1
                  Process: com.tencent.wcdb.sample.encryptdb, PID: 5891
                  java.lang.RuntimeException: An error occurred while executing doInBackground()
                      at android.os.AsyncTask$3.done(AsyncTask.java:318)
                      at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
                      at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
                      at java.util.concurrent.FutureTask.run(FutureTask.java:242)
                      at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
                      at java.lang.Thread.run(Thread.java:761)
                   Caused by: java.lang.IllegalStateException: attempt to re-open an already-closed object: SQLiteDatabase: /data/user/0/com.tencent.wcdb.sample.encryptdb/databases/encrypted.db
                      at com.tencent.wcdb.database.SQLiteClosable.acquireReference(SQLiteClosable.java:56)
                      at com.tencent.wcdb.database.SQLiteDatabase.endTransaction(SQLiteDatabase.java:540)
                      at com.tencent.wcdb.database.SQLiteOpenHelper.getDatabaseLockedLast(SQLiteOpenHelper.java:337)
                      at com.tencent.wcdb.database.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:303)
                      at com.tencent.wcdb.database.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:225)
                      at com.tencent.wcdb.sample.encryptdb.MainActivity$2$1.doInBackground(MainActivity.java:127)
                      at com.tencent.wcdb.sample.encryptdb.MainActivity$2$1.doInBackground(MainActivity.java:110)
                      at android.os.AsyncTask$2.call(AsyncTask.java:304)
                      at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                      at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243) 
                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
                      at java.lang.Thread.run(Thread.java:761) 
D/EGL_emulation: eglMakeCurrent: 0xa3d3efc0: ver 2 0 (tinfo 0xa47a0790)
D/OpenGLRenderer: endAllActiveAnimators on 0x97175f80 (RippleDrawable) with handle 0x971b20b0
Application terminated.
```

**Comment by dotw** [comment]
I confirm the following command causes the error, but I don't know why. 
```
db.execSQL("ATTACH DATABASE ? AS old KEY ?;", new String[]{oldDbFile.getPath(), "passphrase"});
```

**Comment by John-He-928** [comment]
Failing to ATTACH encrypted databases is caused by default page size of SQLCipher, which is hard-coded to 1024 but the actual page size we use is generally 4096.

This is fixed in dev branch by setting default page size of SQLCipher to 4096 on initialization.

You can also fix this issue by executing `PRAGMA cipher_default_page_size=4096;` before your ATTACH statement.

**Comment by John-He-928** [comment]
Fixed on v1.0.6
